src/wal/src/local_storage_impl/segment.rs

```diff
@@ -0,0 +1,741 @@
+
+            if segment_id as i32 > max_segment_id {
+                max_segment_id = segment_id as i32;
+            }
+            all_segments.insert(segment_id, segment_arc);
+        }
+
+        // If no existing segments, create a new one
+        if max_segment_id == -1 {
+            max_segment_id = 0;
+            let path = format!("{}/segment_{}.wal", segment_dir, max_segment_id);
+            let new_segment = Segment::new(path, max_segment_id as u64)?;
+            let segment_arc = Arc::new(Mutex::new(new_segment));
+            all_segments.insert(0, segment_arc);
+        }
+
+        Ok(Self {
+            all_segments: Mutex::new(all_segments),
+            cache: Mutex::new(VecDeque::new()),
+            cache_size,
+            _segment_dir: segment_dir,
+            latest_segment_idx: AtomicU64::new(max_segment_id as u64),
+            log_encoding: CommonLogEncoding::newest(),
+            // todo: do not use MIN_SEQUENCE_NUMBER, read from the latest record
+            next_sequence_num: AtomicU64::new(MIN_SEQUENCE_NUMBER + 1),
+            runtime,
+        })
+    }
+
+    fn get_segment(&self, segment_id: u64) -> Result<Arc<Mutex<Segment>>> {
+        let mut cache = self.cache.lock().unwrap();
+        let all_segments = self.all_segments.lock().unwrap();
+
+        let segment = all_segments.get(&segment_id);
+
+        let segment = match segment {
+            Some(segment_arc) => segment_arc,
+            None => return SegmentNotFound { id: segment_id }.fail(),
+        };
+
+        // Check if segment is already in cache
+        if cache.iter().any(|id| *id == segment_id) {
+            let segment = all_segments.get(&segment_id);
+            return match segment {
+                Some(segment_arc) => Ok(segment_arc.clone()),
+                None => SegmentNotFound { id: segment_id }.fail(),
+            };
+        }
+
+        // If not in cache, load from disk
+        segment.lock().unwrap().open()?;
+
+        // Add to cache
+        if cache.len() == self.cache_size {
+            let evicted_segment_id = cache.pop_front();
+            // TODO: if the evicted segment is being read or written, wait for it to finish
+            if let Some(evicted_segment_id) = evicted_segment_id {
+                let evicted_segment = all_segments.get(&evicted_segment_id);
+                if let Some(evicted_segment) = evicted_segment {
+                    evicted_segment.lock().unwrap().close()?;
+                } else {
+                    return SegmentNotFound {
+                        id: evicted_segment_id,
+                    }
+                    .fail();
+                }
+            }
+        }
+        cache.push_back(segment_id);
+
+        Ok(segment.clone())
+    }
+
+    pub fn write(&self, _ctx: &WriteContext, batch: &LogWriteBatch) -> Result<SequenceNumber> {
+        let segment = self.get_segment(self.latest_segment_idx.load(Ordering::Relaxed))?;
```

Comment:
I found `latest_segment_idx` is only used here, do you forget to update it?

Also use a atomic integer may not a good choice, since there may exists race condition, I suggest use `current: Mutex<Segment>` to keep the latest segment for write.

Reply:
Currently, reading and writing have only been implemented on a single segment file, so the `latest_segment_idx` will not be updated.

Since all segments are stored in all_segments, I think storing just a segment ID here should be sufficient, so I plan to use `current: Mutex<u64>`.